packages/block-editor/src/components/provider/use-block-sync.js

```diff
@@ -186,7 +186,15 @@ export default function useBlockSync( {
 		}
 	}, [ controlledBlocks, clientId ] );
 
+	const isMounted = useRef( false );
+
 	useEffect( () => {
+		// On mount, controlled blocks are already set in the effect above.
+		if ( ! isMounted.current ) {
+			isMounted.current = true;
+			return;
+		}
```

Comment:
Here I'd like to ask what the `isControlled` value actually means: what's the purpose of `areInnerBlocksControlled` and `setHasControlledInnerBlocks`? The `useBlockSync` logic is one of the most intimidating parts of the `block-editor` store, so let's use this opportunity to clarify things.

Some observations from what I see in the code:
1. The only place where `setHasControlledInnerBlocks` is ever called is this very `useBlockSync` hook. And the only place where `isControlled` can go from `true` to `false` is on unmount, when `unsetControlledBlocks` is called. Therefore, why does `useBlockSync` need this `useEffect` hook to detect changes when it it fully in control over these changes? It could do the `pendingChanges = []` and `setControlledBlocks()` calls imperatively, in `unsetControlledBlocks`.
2. `areInnerBlocksControlled` is called inside `useBlockSync`, and then additionally in `getBlocks`, `__experimentalGetPatternTransformItems`, `__unstableHasActiveBlockOverlayActive` and `NavigationMenuContent`. These usages are incomprehensible to me.

Reply:
It was introduced here: #37484

Reply:
"isControlled" means a block container controls its "block list". It has a dedicated "state" for the block list (value and onChange being passed to the InnerBlocks component).

It happens for thing like "reusable blocks", "post content" block... where the block list is in a separate entity than the main "block list" (root block list). 

 1- if you call `getBlocks()` without argument, you're expected to have the block list excluding all controlled inner blocks (for instance the blocks within a reusable block shouldn't be there)
 2- if you call `getBlocks( rootClientId )` and `rootClientId` is a controlled container (like a wp/block instance), you expect to get the inner blocks of that particular block (the ones excluded from 1)

I hope this helps.

Reply:
Now that I think about it, we should probably add this as "documentation" somewhere.

Reply:
I had to read this three times to get a rough grasp of what was going on. I agree that some documentation with background to clarify why and in which cases this is necessary would be useful. 

I wonder if there's a better way to express in code that the mounted check should apply only to controlled blocks.

Reply:
So, this is how I understand it all works, in my own words. Please confirm or correct:

When blocks control their children, all the blocks are still part of one big `block-editor` tree, is that right? Or are there separate `block-editor` stores, created by context providers and nested registries?

When `ControlledInnerBlocks` mounts, it will start syncing the `clientId` subtree it manages, according to the `value`, and `onChange` props. When it unmounts, it stops syncing.

The controlled, synced blocks are stored in the `block-editor`'s `state.tree` with `controlled||` prefixes. It's basically as if there was a `controlled` branch on that part of the tree, overriding the "original" part of the tree, which is still there in `state.tree`, but stored with unprefixed `clientId` keys.

The `setHasControlledInnerBlocks` action marks a part of tree as controlled. Most directly it affects `getBlocks`, telling it which "branch" it should look at when returning inner blocks of a particular `clientId`. Then there is a handful of other places that check if a given `clientId` subtree is controlled.

From all this, it seems to me that the `useBlockSync` hook shouldn't read the `isControlled` value at all. And shouldn't have an `useEffect( () => { ... }, [ isControlled ] )`. Because the hook _knows_ the value, it knows that the given `clientId` is controlled for the entire lifetime of the hook. That's what the hook is for! It sets `setHasControlledInnerBlocks` to `true` on mount, then does the syncing, and then sets `setHasControlledInnerBlocks` back to `false` when it unmounts.

Reply:
I think the whole `useBlockSync` only applies to controlled blocks? We never use that hooks for non controlled blocks.

Reply:
I agree -- why then does the hook read `isControlled` and does a `setControlledBlocks` call when it goes from `true` to `false`? It should to that call in its cleanup/unmount handler instead. Because that's the only place where `isControlled` can ever change from `true` to `false`.

Reply:
> From all this, it seems to me that the useBlockSync hook shouldn't read the isControlled value at all. And shouldn't have an useEffect( () => { ... }, [ isControlled ] ). Because the hook knows the value, it knows that the given clientId is controlled for the entire lifetime of the hook. That's what the hook is for! It sets setHasControlledInnerBlocks to true on mount, then does the syncing, and then sets setHasControlledInnerBlocks back to false when it unmounts.

Your own words are correct, but I'm not 100% sure about this part but it's not that important.

> I think the whole useBlockSync only applies to controlled blocks? We never use that hooks for non controlled blocks.

It might not be the case anymore, it used to be called for all InnerBlocks component.